providers/azure_provider/azure.go

```diff
@@ -153,17 +186,191 @@ func (ac *AzureConfigure) GetAzureClusterContext(groupName, clusterName string)
+	}
+
+	update, err := ac.GroupClient.CreateOrUpdate(ac.Ctx, tenant+"-"+util.SopsRootName, params, nil)
+	if err != nil {
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == 403 {
+			zap.S().Warnf("permissions denied to create Azure resource group: %s", tenant+"-"+util.SopsRootName)
+			return nil
+		}
+
+		return err
+	}
+
+	ac.ResourceGroupName = *update.Name
+
+	zap.S().Infof("created Azure resource group: %s", ac.ResourceGroupName)
+
+	return nil
+}
+
+func (ac *AzureConfigure) existsKeyVaultResourceGroup(tenant string) (bool, error) {
+	existence, err := ac.GroupClient.CheckExistence(ac.Ctx, tenant+"-"+util.SopsRootName, nil)
+	if err != nil {
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == 403 {
+			zap.S().Warnf("permissions denied to check existence Azure resource group: %s",
+				tenant+"-"+util.SopsRootName)
+			return false, nil
+		}
+
+		return false, err
+	}
+
+	return existence.Success, nil
+}
+
+func (ac *AzureConfigure) CreateAzureKeyVault(tenant string) error {
+	if err := ac.createKeyVaultResourceGroup(tenant); err != nil {
+		return err
+	}
+
+	if len(ac.ResourceGroupName) == 0 {
+		ac.ResourceGroupName = tenant + "-" + util.SopsRootName
+	}
+
+	ac.KeyVaultName = generateKeyVaultName(tenant)
+
+	params := armkeyvault.VaultCreateOrUpdateParameters{
+		Location: to.Ptr(ac.Location),
+		Properties: &armkeyvault.VaultProperties{
+			EnableRbacAuthorization: to.Ptr(true),
+			SKU: &armkeyvault.SKU{
+				Family: to.Ptr(armkeyvault.SKUFamilyA),
+				Name:   to.Ptr(armkeyvault.SKUNameStandard),
+			},
+			TenantID: to.Ptr(ac.TenantID),
+		},
+	}
+
+	update, err := ac.VaultsClient.BeginCreateOrUpdate(ac.Ctx, ac.ResourceGroupName, ac.KeyVaultName, params, nil)
+	if err != nil {
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == 403 {
+			zap.S().Warnf("permissions denied to create Azure key vault: %s",
+				ac.KeyVaultName)
+			return nil
+		}
+
+		return err
+	}
+
+	result, err := update.PollUntilDone(ac.Ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	ac.KeyVaultURI = *result.Properties.VaultURI
+
+	zap.S().Infof("created Azure key vault: %s, %s", ac.KeyVaultName, ac.KeyVaultURI)
+
+	return nil
+}
+
+func (ac *AzureConfigure) GetAzureKeyVault(tenant string) (bool, error) {
+	if ok, err := ac.existsKeyVaultResourceGroup(tenant); err != nil {
+		return false, err
+	} else if !ok {
+		return false, nil
+	}
+
+	ac.KeyVaultName = generateKeyVaultName(tenant)
+	ac.ResourceGroupName = tenant + "-" + util.SopsRootName
+
+	resp, err := ac.VaultsClient.Get(ac.Ctx, tenant+"-"+util.SopsRootName, ac.KeyVaultName, nil)
+	if err != nil {
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == 404 {
+			return false, nil
+		}
+
+		if errors.As(err, &respErr) && respErr.StatusCode == 403 {
+			zap.S().Warnf("permissions denied to get Azure Key Vault: %s", ac.KeyVaultName)
```

Comment:
seems that we should fixed the naming in every log line to be: "Azure **K**ey **V**ault"
some places have: "created Azure **k**ey **v**ault"

Reply:
key vault is not a common noun